rclcpp/include/rclcpp/strategies/message_pool_memory_strategy.hpp

```diff
@@ -52,11 +59,28 @@ class MessagePoolMemoryStrategy
 
   /// Default constructor
   MessagePoolMemoryStrategy()
-  : next_array_index_(0)
   {
     for (size_t i = 0; i < Size; ++i) {
-      pool_[i].msg_ptr_ = std::make_shared<MessageT>();
-      pool_[i].used = false;
+      pool_[i] = static_cast<MessageT *>(malloc(sizeof(MessageT)));
+      free_list_.push_back(i);
+    }
+  }
+
+  ~MessagePoolMemoryStrategy()
+  {
+    // The user may have held onto shared pointers after a borrow_message().  In that case,
+    // freeing the memory from the pool may lead to UB.  If we detect the situation where this
+    // class is being destroyed before the shared pointers, warn the user.
+
+    if (free_list_.size() != Size) {
+      RCLCPP_WARN(
+        rclcpp::get_logger("MessagePool"),
+        "User code is holding onto shared pointers from the message pool; this will leak memory");
+    }
```

Comment:
This is fine for now (or forever if no one cares to improve it), but I just wanted to note that we could probably do something so that the destructor of the shared pointers will free the memory if the pool no longer exists. Maybe that could be a todo or something. This improvement would also let us optimize a bit and maybe allocate/free the memory all at once, maybe by having all the shared pointers that are borrowed also have a shared pointer to the arena of memory they're allocated from, that way after the last one is destroyed the actual memory for the pool would also be deleted (assuming the memory pool had let its handle on it go too).

Either way, it might be worth mentioning this limitation / expected pre condition to destroying the pool in the pool's docstring.

Reply:
I can't suggest on the docstring because it's not part of the diff, but something like:

> All borrowed message must be returned before deleting the memory pool otherwise memory will be leaked.